server/src/unifyfs_server_pid.c

```diff
@@ -0,0 +1,207 @@
+
+    out.ret = 0;
+    hret = margo_respond(handle, &out);
+    if (hret != HG_SUCCESS) {
+        LOGERR("failed to respond rpc (ret=%d)", hret);
+        return;
+    }
+
+    margo_free_input(handle, &in);
+    margo_destroy(handle);
+
+    for (i = 0; i < glb_pmi_size; i++) {
+        if (server_pids[i] > 0) {
+            count++;
+        }
+    }
+
+    if (count == glb_pmi_size) {
+        ret = pthread_cond_signal(&server_pid_cond);
+        if (ret) {
+            LOGERR("failed to signal condition (%s)", strerror(ret));
+        }
+    }
+}
+DEFINE_MARGO_RPC_HANDLER(server_pid_handle_rpc);
+
+static int set_timeout(void)
+{
+    int ret = 0;
+    long timeout_sec = 0;
+
+    if (server_cfg.server_init_timeout) {
+        ret = configurator_int_val(server_cfg.server_init_timeout,
+                                   &timeout_sec);
+        if (ret) {
+            LOGERR("failed to read configuration");
+            return ret;
+        }
+    }
+
+    clock_gettime(CLOCK_REALTIME, &server_pid_timeout);
+    server_pid_timeout.tv_sec += timeout_sec;
+
+    return 0;
+}
+
+static int create_server_pid_file(void)
+{
+    int i = 0;
+    int ret = 0;
+    char filename[PATH_MAX] = { 0, };
+    FILE* fp = NULL;
+
+    sprintf(filename, "%s/%s", server_cfg.sharedfs_dir, UNIFYFSD_PID_FILENAME);
+
+    fp = fopen(filename, "w");
+    if (!fp) {
+        LOGERR("failed to create file %s (%s)", filename, strerror(errno));
+        ret = errno;
+        goto out;
+    }
+
+    for (i = 0; i < glb_pmi_size; i++) {
+        fprintf(fp, "[%d] %d\n", i, server_pids[i]);
+    }
+
+    fclose(fp);
+
+out:
+    return ret;
+}
+
+int unifyfs_publish_server_pids(void)
+{
+    int ret = UNIFYFS_SUCCESS;
+
+    if (glb_pmi_rank > 0) {
+        ret = server_pid_invoke_rpc();
+        if (ret) {
+            LOGERR("failed to invoke pid rpc (%s)", strerror(ret));
+        }
+    } else {
+        ret = set_timeout();
+        if (ret) {
+            goto out;
+        }
+
+        server_pids = calloc(glb_pmi_size, sizeof(*server_pids));
+        if (!server_pids) {
+            LOGERR("failed to allocate memory (%s)", strerror(errno));
+            goto out;
+        }
+
+        server_pids[0] = server_pid;
+
+        if (glb_pmi_size > 1) {
+            ret = pthread_cond_timedwait(&server_pid_cond,
+                                         &server_pid_mutex,
+                                         &server_pid_timeout);
+            if (ETIMEDOUT == ret) {
+                LOGERR("some servers failed to initialize within timeout");
+                goto out_free;
+            } else if (ret) {
+                LOGERR("failed to wait on condition (err=%d, %s)",
+                       errno, strerror(errno));
+                goto out_free;
+            }
+        }
+
+        ret = create_server_pid_file();
+        if (UNIFYFS_SUCCESS == ret) {
+            LOGDBG("all servers are ready to accept client connection");
+        }
+
+out_free:
```

Comment:
instead of having two labels at the end of this function, just use one like this:
```
out:
    if (NULL != server_pids) {
        free(server_pids);
    }
    return  ret;
```